src/voter_set.rs

```diff
@@ -12,148 +12,241 @@
 #[derive(Clone, PartialEq, Eq)]
 #[cfg_attr(feature = "std", derive(Debug))]
 pub struct VoterSet<Id: Hash + Eq> {
-	weights: HashMap<Id, VoterInfo>,
-	voters: Vec<(Id, u64)>,
-	threshold: u64,
+	/// The voters in the set.
+	voters: HashMap<Id, VoterInfo>,
+	/// The total order associated with the keys in `voters`.
+	order: Vec<Id>,
+	/// The required weight threshold for supermajority w.r.t. this set.
+	threshold: VoterWeight,
+	/// The total weight of all voters.
+	total_weight: VoterWeight,
 }
 
 impl<Id: Hash + Eq> VoterSet<Id> {
-	/// Get the voter info for a voter.
-	pub fn info<'a>(&'a self, id: &Id) -> Option<&'a VoterInfo> {
-		self.weights.get(id)
-	}
+	/// Create a voter set from a weight distribution produced by the given iterator.
+	///
+	/// If the distribution contains multiple weights for the same voter ID, they are
+	/// understood to be partial weights and are accumulated. As a result, the
+	/// order in which the iterator produces the weights is irrelevant.
+	///
+	/// Returns `None` if the iterator produced no non-zero weights, i.e.
+	/// the voter set would be empty.
+	///
+	/// # Panics
+	///
+	/// If the total voter weight exceeds `u64::MAX`.
+	///
+	pub fn new<I>(weights: I) -> Option<Self>
+	where
+		Id: Ord + Clone,
+		I: IntoIterator<Item = (Id, u64)>
+	{
+		let weights = weights.into_iter();
+
+		// Populate the voter set, thereby calculating the total weight.
+		let mut voters = HashMap::with_capacity(weights.size_hint().0);
+		let mut total_weight = 0u64;
+		for (id, weight) in weights {
+			if let Some(w) = NonZeroU64::new(weight) {
+				// Prevent construction of inconsistent voter sets by checking
+				// for weight overflow (not just in debug mode). The protocol
+				// should never run with such voter sets.
+				total_weight = total_weight.checked_add(weight).expect("Voter weight overflow");
+				match voters.entry(id) {
+					Entry::Vacant(e) => {
+						e.insert(VoterInfo {
+							position: 0, // The total order is determined afterwards.
+							weight: VoterWeight(w)
+						});
+					},
+					Entry::Occupied(mut e) => {
+						let v = e.get_mut();
+						let n = v.weight.get() + weight;
+						let w = NonZeroU64::new(n).expect("nonzero + nonzero is nonzero");
+						v.weight = VoterWeight(w);
+					}
+				}
+			}
+		}
 
-	/// Get the length of the set.
-	pub fn len(&self) -> usize { self.voters.len() }
+		if voters.is_empty() {
+			// No non-zero weights; the set would be empty.
+			return None
+		}
 
-	/// Whether the set contains the key.
-	pub fn contains_key(&self, id: &Id) -> bool {
-		self.weights.contains_key(id)
-	}
+		let total_weight = VoterWeight::new(total_weight).expect("voters nonempty; qed");
 
-	// Get voter by index.
-	pub fn voter_by_index(&self, idx: usize) -> &(Id, u64) {
-		&self.voters[idx]
-	}
+		// Establish the total order on the set based on the voter IDs.
+		let mut order = voters.keys().cloned().collect::<Vec<_>>();
+		order.sort_unstable();
```

Comment:
All nodes need to agree on the canonical ordering (for unanimous primary selection), so a stable sort is safest. I don't think different versions of `core` are required to use the same unstable sorting algorithm, which could lead to consensus confusion.

Reply:
But the difference between a stable and unstable sort only related to equal elements. There are no two equal elements in `order` by its definition from the line above. What am I missing?